### PR TITLE
T/52: Code

### LINKED ISSUE #52: Inline Code
There should be support for inline code markup, outputting the `<code>` HTML element.

### PR DESCRIPTION
**Pull request by pomek**
### Suggested merge commit message ([convention](https://github.com/ckeditor/ckeditor5-design/wiki/Git-commit-message-convention))

Feature: Introduced the `Code` plugin. Closes #52.

---

### Additional information

Autoformatting will be done in https://github.com/ckeditor/ckeditor5-autoformat/issues/35.

### CODE REVIEW

**Comment by szymonkups** [comment]
There are some problems with single space handling:
* create empty paragraph, change to code style, try to enter space - nothing happen
* in non-empty paragraph create code at the end of it type some letters and try to type space

It looks like space handling at the beggining/end of the code element is broken.

**Comment by pomek** [comment]
I tried to change the name of the element in view (tested on `div`, `span`, `tt` and even with custom elements `_code` and `code_`) and everything works fine.

IMO it looks like the `code` element has some super attributes and the editor does not render the spaces at the end of the element. In the model, these spaces are adding properly.

**Comment by Reinmar** [comment]
This problem with spaces deserves a separate ticket in ckeditor5-engine. And should not block this ticket.

**Comment by pomek** [comment]
Reported in https://github.com/ckeditor/ckeditor5-engine/issues/1116. So, could we merge this PR?

**Comment by pomek** [comment]
I added some styles but I am not CSS master, so it may require some changes.

File: theme/theme.scss [hunk 1]
```diff
@@ -0,0 +1,27 @@
+// Copyright (c) 2003-2017, CKSource - Frederico Knabben. All rights reserved.
```
Comment (Reinmar) [inline theme/theme.scss]: @oleq, do you think we could ship with such code styles? They fix `<code>` rendering (GH and Umberto use similar ones).

File: theme/theme.scss [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+// For licensing, see LICENSE.md or http://ckeditor.com/license
+
+@import '~@ckeditor/ckeditor5-theme-lark/theme/helpers/_colors.scss';
+@import '~@ckeditor/ckeditor5-theme-lark/theme/helpers/_rounded.scss';
+@import '~@ckeditor/ckeditor5-theme-lark/theme/helpers/_spacing.scss';
+
+// It's a darker shade of u-color( 'background-hue' ) so it looks good in info-boxes.
```
Comment (Reinmar) [inline theme/theme.scss]: Like... what?

File: theme/theme.scss [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+	border-radius: $ck-border-radius;
+
+	&:before,
+	&:after {
+		// To reduce the width of that space to 0px.
+		letter-spacing: -1 * ck-spacing();
+		content: "\00a0";
+	}
+
+	// To remove an empty line when `code` is empty.
```
Comment (Reinmar) [inline theme/theme.scss]: This comment is unclear. This line is needed to fix empty code rendering in empty block.

But it was a good catch :)

**Comment by Reinmar** [comment]
I'll fix the those two issues that I found in the theme.scss.

**Comment by Reinmar** [comment]
The styling of the code element (the CSS generated spaces), unfortunately, add another selection step in Firefox:

sep-06-2017 14-22-18

OTOH, it works pretty well with it.

**Comment by Reinmar** [comment]
On Safari an error is thrown when typing spaces at the end of the `<code>` element. Might be related to https://github.com/ckeditor/ckeditor5-engine/issues/1116, so I'll add there a note.

**Comment by Reinmar** [comment]
@oleq, could you take a look at that SASS file and on how it affects FF?

**Comment by Reinmar** [comment]
Hm, I have a better idea, actually:

1. let's remove those styles – to KISS,
2. this means that spaces at the boundaries of `<code>` will be completely lost which leads to editor throwin an error if you type multiple of them:

    ```
    input.js:294 Uncaught TypeError: Cannot read property 'index' of undefined
        at MutationHandler._handleTextNodeInsertion (input.js:294)
        at MutationHandler.handle (input.js:152)
        at Input._handleMutations (input.js:105)
        at Document.listenTo (input.js:50)
        at Document.fire (emittermixin.js:281)
        at MutationObserver._onMutations (mutationobserver.js:248)
    ```

    I don't know why this error is actually throw. I suppose that it may have something to do with https://github.com/ckeditor/ckeditor5-engine/issues/1125 because I saw similar stack error there.

3. But the way to prevent that is to remove `code` from here:

   https://github.com/ckeditor/ckeditor5-engine/blob/0bc1097d0448c13bf7e7480ceafb5e738d1eb93f/src/view/domconverter.js#L68

   After doing this, multiple spaces in `<code>` elements will be rendered as a mix of normal spaces and non-breaking spaces.

   sep-06-2017 14-40-32

**Comment by Reinmar** [comment]
To make it clear, with the proposed changes, such data will be generated:

```html
<p>Fo <code>o &nbsp; b &nbsp; bom &nbsp; </code>ar</p>
```

Is it ok?

**Comment by fredck** [comment]
I think that not only this is ok but it is also the very right way to do it, in the HTML point of view.

I was just wondering what happens if one copies and pastes this code into real code... would nbsp characters end up there?

**Comment by Reinmar** [comment]
> I was just wondering what happens if one copies and pastes this code into real code... would nbsp characters end up there?

In the model you'll have normal spaces. In the DOM you'll have the same as I showed above. Basically – on DOM -> view conversion (upon pasting) unnecessary `&nbsp;`s will be turned to normal spaces.

**Comment by fredck** [comment]
> In the model you'll have normal spaces. In the DOM you'll have the same as I showed above. Basically – on DOM -> view conversion (upon pasting) unnecessary &nbsp;s will be turned to normal spaces.

This doesn't help much, because one will still have to render the output HTML inside browsers... and the only way to do that is doing exactly how we do it.

**Comment by Reinmar** [comment]
> This doesn't help much, because one will still have to render the output HTML inside browsers... and the only way to do that is doing exactly how we do it.

??? But you get data through `getData()` and see this:

```html
<p><i>This</i> <code>is &nbsp; &nbsp; an</code> <strong>editor</strong> <u>ins</u><code>is &nbsp; &nbsp; an</code><u>tance</u>.</p>
```

Which renders correctly.

**Comment by fredck** [comment]
Actually, what HTML the editor itself would output for this?

**Comment by fredck** [comment]
Ok then... so we have to confirm that copy and paste from that produces "pure" spaces.

**Comment by Reinmar** [comment]
@fredck please see my above comments ;) I confirmed exactly that:

https://github.com/ckeditor/ckeditor5-basic-styles/pull/55#issuecomment-327473638

And also what data will be returned:

https://github.com/ckeditor/ckeditor5-basic-styles/pull/55#issuecomment-327472093

**Comment by fredck** [comment]
That's wonderful then... so your proposal is the very right way for it ;)

**Comment by fredck** [comment]
To note that other converters, like Markdown, will have a sequence of plain spaces there instead. It'll be the job of the third-party MD-to-HTML libraries to convert those spaces on something that renders right inside browsers.

**Comment by Reinmar** [comment]
@pomek – let's close https://github.com/ckeditor/ckeditor5-engine/issues/1126 first. And then, let's clean up this PR.

**Comment by Reinmar** [comment]
> To note that other converters, like Markdown, will have a sequence of plain spaces there instead. It'll be the job of the third-party MD-to-HTML libraries to convert those spaces on something that renders right inside browsers.

Haha, it's not like that. GFM e.g. leaves spaces as they are and GH or other websites using it need to fix this using that hacky CSS I posted in https://github.com/ckeditor/ckeditor5-engine/issues/1116#issuecomment-326243007. We've been there with Umberto too ;<

**Comment by fredck** [comment]
> Haha, it's not like that. GFM e.g. leaves spaces as they are and GH or other websites using it need to fix this using that hacky CSS I posted in ckeditor/ckeditor5-engine#1116 (comment). We've been there with Umberto too ;<

Yes, ofc. What I meant to say is that systems that use CKEditor 5 to output Markdown will have to implement their MD-to-HTML scripts in a way that spaces are properly converted IF they want to have browsers rendering their data in the way their users see it in the editor.

This means that we can already predict that many systems will be incompatible with that and that a plugin that avoids multiple spaces will be born to "solve" this problem, the other way around.

**Comment by pomek** [comment]
@Reinmar, after merging changes in engine, this PR is ready to review once again.

**Pull request merged**